### PR TITLE
Foxglove Support

### PR DESCRIPTION
**Pull request by pfeiferj**
#30272 

This takes a different approach than the demo mentioned in the issue. Instead of storing the raw capnp messages in the mcap file this converts the data to json to store inside of the mcap file. This allows using the data in foxglove without having to maintain a fork of the studio.

Features:
* Downloads segments/routes from comma connect
* Generates json schema for foxglove consumption from cereal capnp definitions
* Converts logs to mcap (json formatted messages) with included schema, mcap file can be loaded in the foxglove  desktop app or web app
* Some messages are copied into well-known foxglove schema types:
  * thumbnail -> CompressedImage
  * errorLogMessage -> Log
  * logMessage -> Log
  * modelV2.temporalPose -> FrameTransform
  * liveLocationKalman -> LocationFix
* Can optionally embed fcam, dcam, and ecam

Screenshot 2023-12-28 at 10 52 08 AM

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
> * concatenate multiple log files into mcap file

Can it work with segments? The concatenation is pretty annoying to have to do.

**Comment by pfeiferj** [comment]
> > * concatenate multiple log files into mcap file
> 
> Can it work with segments? The concatenation is pretty annoying to have to do.

I believe that foxglove only supports loading a single file at a time, so if you want to view multiple segments at once i think you would have to concatenate them. Probably should explicitly support a single segment though

**Comment by pfeiferj** [comment]
@adeebshihadeh This is probably good to go. There's definitely still things that could be added/improved upon but it should be a decent base

File: tools/foxglove/cereal_to_mcap.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python3
```
Comment (adeebshihadeh) [inline tools/foxglove/cereal_to_mcap.py:1]: this file isn't actually executable

File: tools/foxglove/cereal_to_mcap.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+import cereal
```
Comment (adeebshihadeh) [inline tools/foxglove/cereal_to_mcap.py:3]: imports should be sorted by std lib and third party, newline, then openpiiot imports

File: tools/foxglove/cereal_to_mcap.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+from openpilot.selfdrive.test.openpilotci import get_url
+from openpilot.tools.foxglove.json_schema import get_event_schemas
+from openpilot.tools.foxglove.foxglove_schemas import RAW_IMAGE, COMPRESSED_IMAGE, FRAME_TRANSFORM, LOCATION_FIX, LOG
+from openpilot.tools.foxglove.utils import register_schema, register_channel, register, message
+from openpilot.tools.lib.framereader import FrameReader
+from openpilot.tools.foxglove.transforms import transform_camera, TRANSFORMERS
+
+foxglove_dir = os.path.dirname(os.path.realpath(__file__))
+
+DEMO_ROUTE = "a2a0ccea32023010|2023-07-27--13-01-19"
```
Comment (adeebshihadeh) [inline tools/foxglove/cereal_to_mcap.py:23]: import this from plotjuggler

File: tools/foxglove/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+Examples using segment name:
+
+`./cereal_to_mcap.py "a2a0ccea32023010|2023-07-27--13-01-19--1"`
+
+## Demo
+
+For a quick demo, go through the installation step and run this command:
+
+`./cereal_to_mcap.py --demo --qlog`
```
Comment (adeebshihadeh) [inline tools/foxglove/README.md:45]: this didn't work on my machine
```
(openpilot-py3.11) bash-3.2$ ./cereal_to_mcap.py --demo --qlog
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
Loading Segment(s)


warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points



Segment(s) Loaded
Starting Conversion
Percent converted: 0.07%Traceback (most recent call last):
  File "/Users/adeebshihadeh/openpilot/tools/foxglove/./cereal_to_mcap.py", line 199, in <module>
    download_and_convert_route(route_or_segment_name, args.segment_count, args.qlog, args.ci, args.fcam, args.dcam, args.ecam)
  File "/Users/adeebshihadeh/openpilot/tools/foxglove/./cereal_to_mcap.py", line 94, in download_and_convert_route
    convert_log(f"{str(route_or_segment_name.canonical_name).split('|')[1]}.mcap", tmp.name, cams)
  File "/Users/adeebshihadeh/openpilot/tools/foxglove/./cereal_to_mcap.py", line 172, in convert_log
    message(writer, typeToChannel[str(event.which)], e, offset, e)
  File "/Users/adeebshihadeh/openpilot/openpilot/tools/foxglove/utils.py", line 26, in message
    writer.add_message(
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/writer.py", line 196, in add_message
    self.__chunk_builder.add_message(message)
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/_chunk_builder.py", line 45, in add_message
    message.write(self.record_writer)
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/records.py", line 312, in write
    stream.write8(self.log_time)
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/data_stream.py", line 104, in write8
    self.write(struct.pack("<Q", value))
               ^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: argument out of range
(openpilot-py3.11) bash-3.2$ 
```

**Review by adeebshihadeh — COMMENTED** [review]
Only did a quick review since it didn't work for me. Are all these lines necessary?

I like that this doesn't require a foxglove fork, but I'm concerned about the performance of this, particularly with 50+ segments. Got any quick benchmarks of routes of various sizes?

File: tools/foxglove/cereal_to_mcap.py (hunk 1 above)
Comment (pfeiferj) [reply tools/foxglove/cereal_to_mcap.py:1]: interesting, it has the execute bit set on a checkout for me and if i remove the execute bit it shows up as a diff in git

File: tools/foxglove/README.md (hunk 4 above)
Comment (pfeiferj) [reply tools/foxglove/README.md:45]: i just fixed the demo route. had to do with me calculating the timestamp using the difference between wallTimeMonos and the logMonoTime. I'm just ignoring the offset now if wallTimeMonos is 0, but that does result in the demo route having incorrect timestamps in foxglove.

What I don't understand is why the warning logs repeated like that. The part of code that prints those warnings out only runs once at the beginning of the script. Once it starts loading the segments it shouldn't be possible to see those warnings printed out anymore. It's almost like you had multiple instances of the script running

**Comment by pfeiferj** [comment]
> Only did a quick review since it didn't work for me. Are all these lines necessary?
I suppose that depends on what you want out of it.

A lot of the code is for transforming some of the openpilot messages into foxglove types for better integration into foxglove. For example, taking the liveLocationKalman and putting it into the format that enables the map view in foxglove. If you want to have that functionality you would have to have code for doing that transformation regardless of if you fork foxglove or not.

There's also a decent amount of code that's just examples of "user scripts" for converting data from within foxglove. Those are purely examples so aren't necessary.

some line counts for less necessary things:
transforming data: 139
schemas for transforming data: 314
example code: 126


> I like that this doesn't require a foxglove fork, but I'm concerned about the performance of this, particularly with 50+ segments. Got any quick benchmarks of routes of various sizes?

performance is definitely not stellar. 40 segments of rlogs is on the order of 10 minutes on my machine to convert. this type of script is unfortunately one of the worst case scenarios for python. Always amazes me how bad python is at loops. Probably could speed it up some by completely removing binary data from the mcap file, such as the thumbnails and can data. Converting the binary data to base64 is fairly expensive. There's also a fair bit of slowdown due to foxglove not supporting NaN in the json, but i'm not sure there's a good solution for that.

It's also worth noting that the way the example project works also required looping through every event in the logs and re-writing it into the mcap file so i'd guess the difference in speed would be like 2-3x instead of an order of magnitude slower

**Comment by pfeiferj** [comment]
> performance is definitely not stellar. 40 segments of rlogs is on the order of 10 minutes on my machine to convert. this type of script is unfortunately one of the worst case scenarios for python. Always amazes me how bad python is at loops. Probably could speed it up some by completely removing binary data from the mcap file, such as the thumbnails and can data. Converting the binary data to base64 is fairly expensive. There's also a fair bit of slowdown due to foxglove not supporting NaN in the json, but i'm not sure there's a good solution for that.

potentially there could be a pretty good speed up by going to flatbuffers instead of json. Some of the biggest performance impacts (the base64 and the nan->None conversions) _should_ go away with flatbuffers

**Comment by adeebshihadeh** [comment]
Alright, it's working for me now. So this is a good start, but before we merge, we need either 1) the conversion to be fast or 2) the experience to be more complete.

By more complete, I mean it all just "works" albeit slowly. You can run `./foxglove.py --demo` and it launches foxglove with a nice starter layout for openpilot debugging (map, qcam, and some plots?).

**Comment by adeebshihadeh** [comment]
@pfeiferj FYI, this bounty will be unlocked soon unless activity starts up again (as stated in the bounty docs).

**Pull request closed without merging**